Address Discovery and Advertisement (#89)
libp2p/docs: content/concepts/address-discovery-and-advertisement.md

    - Note that the connection with the Relay server should be a long lived one as peers can ONLY connect to it
    using the relay address if it is connected to the Relay server.

    - If a peer discovers it has public reachability by seeing an “EvtLocalReachabilityChanged” event for it,
    it stops advertising Relay addresses.

    - When a peer with private reachability loses connection to it’s Relay server, it will remove the corresponding
    relay addresses from it’s address set and look for a new Relay server to connect to.

    - When a peer connects to a new Relay server, it adds the corresponding Relay address to it’s set
    of dialable addresses.

    - The component that listens to NAT reachability events, connects to Relay servers and builds/advertises
    Relay addresses is called [AutoRelay][repo_autorelay]. It can be enabled by using the `libp2p.EnableRelay(nil)` and `libp2p.EnableAutoRelay()` option on the Host.

{{% notice "note" %}}
We can also configure AutoRelay to use static relay servers rather than discovering them via DHT by
using the `libp2p.StaticRelays(relayServerAddrs...)` option on the Host.
{{% /notice %}}

### Advertising Dialable Addresses

> The first section covered how a libp2p nodes “discovers” it’s dialable addresses.
> All the methods and address sets discussed above go into making the final set of a node’s dialable addresses.
> The way a libp2p peer shares these dialable addresses with other peers is:

Reviewer: ```suggestion
The first section covered how a libp2p nodes "discovers" its dialable addresses. All the methods and address sets discussed above go into making the final set of a node’s advertised addresses. The way a libp2p peer shares these dialable addresses with other peers is as follows:
```

> - When a peer connects to other peers, it asks it’s `Host` for it’s current set of dialable addresses.
> 
> - This will include some or all of the address sets discussed above (Interface, Relay, etc. etc.).

Reviewer: ```suggestion
- When a peer connects to other peers, they exchange their dialable addresses via the `Identify` protocol.
    - This will include some or all of the address sets discussed above (Interface, Relay, etc. etc.).
```

> - Put those addresses in the `listenAddrs` field of the `Identify` message and send the `Identify` message to
>   the other peer.
> 

Reviewer: This is confusing and is now covered by the previous point. I don't think we need to specify the protobuf field.
```suggestion
```

> - When a peer receives another peer’s dialable addresses from the `listenAddrs` field of the `Identify` message,
>   it stores them in it’s [peerstore][peerstore_repo] for some time. The timing does NOT matter for this discussion.

Reviewer: Just remove the timing mention since we're not talking about it.
```suggestion
- When a peer receives another peer’s dialable addresses from the `Identify` exchange it stores them in its [peerstore][peerstore_repo].
```

> - Any peer can then lookup a peer’s addresses by asking the DHT network to find the peer with that [peerID][peer_id] by
>   using the [FindPeer][find_peer] DHT API. The DHT nodes will use the addresses stored in their peerstore as described above
>   to find and return the addresses of the peer.

Reviewer: We don't mention how this actually succeeds as there is no mention of me connecting to my closest peers. The lookup won't work if I dont regularly connect to my closest peers.
```suggestion
- In order to allow other peers to discover our addresses, libp2p will intermittently query the DHT for its closest peers and connect to them. This will result in addresses being exchanged and stored with those peers, so that future DHT [FindPeer][find_peer] queries searching for our [peerID][peer_id] will discover our addresses.
```

>     #### Handling changes to the peer’s dialable addresses

Reviewer: ```suggestion
#### Handling changes to the peer’s dialable addresses
```

>     All well and good so far. But the set of a peer’s dialable addresses is ofcourse not a static set. A peer can
>     start listening on new interfaces, see more of it’s observed addresses, change it’s network, be assigned a new
>     IP address by the Router, etc. Hence, we also need a mechanism to detect changes to a peer’s dialable addresses set
>     and advertise these changes to the world on the fly. The way we do this is by
>     **running an event loop** in the `Host`.

Reviewer: Avoid language like "of course". It may not be obvious to the reader (new to libp2p) and it can be demeaning in those instances.

```suggestion
All well and good so far, but the set of a peer’s dialable addresses is not a static set. A peer can start listening on new interfaces, see more of its observed addresses, change its network, be assigned a new IP address by the Router or ISP, and more. Hence, we also need a mechanism to detect changes to a peer’s dialable addresses set and advertise these changes to the world as they happen. In go-libp2p we do this is by **running an event loop** in the `Host`.
```

>     #### When does the Host event loop fire ?

Reviewer: ```suggestion
##### When does the Host event loop fire ?
```

>     - Every 5 seconds.
> 
>     - When the user starts listening on a new interface address/es by calling `Host.Network.Listen(addrs...)`.
> 
>     - When `AutoRelay` detects a change in it’s reachability(private -> public or vice versa), creates a
>     connection with a new relay server or loses connection with an existing relay server since all of these would
>     cause a change in it’s relay addresses.
> 
>     #### What does the Host do when the event loop fires ?
>     - Updates it’s primary network interface address by using the netroute library as discussed above
>     (if it has changed). This takes care of network changes/routers assigning new addresses via DHCP.
> 
>     - Builds a new address set by using all the addresses discovered by the various mechanisms described in the
>     `Address Discovery` section and compares it with the previous address set.
> 
>     - If the new address set is different from the previous address set, it fires an `EvtLocalAddressesUpdated` event.
> 
>     - The ID service listens for the `EvtLocalAddressesUpdated` event and pushes the new set of addresses to all
>     peers it is connected to via the `Identify Push` protocol.
>     The ID service is the same service that also handles the `Identify` protocol described above.

Reviewer: ```suggestion
- On interval, every 5 seconds.

- When the user starts listening on a new interface address/es by calling `Host.Network.Listen(addrs...)`.

- When `AutoRelay` detects a change in it’s reachability(private -> public or vice versa), creates a connection with a new relay server, or loses connection with an existing relay server since all of these would cause a change in its relay addresses.

##### What does the Host do when the event loop fires ?
- Updates its primary network interface address by using the netroute library as discussed above (if it has changed). This takes care of network changes/routers assigning new addresses via DHCP.

- Builds a new address set by using all the addresses discovered by the various mechanisms described in the `Address Discovery` section and compares it with the previous address set.

- If the new address set is different from the previous address set, it fires an `EvtLocalAddressesUpdated` event.
    - The ID service listens for the `EvtLocalAddressesUpdated` event and pushes the new set of addresses to all peers it is connected to via the `Identify Push` protocol. The ID service is the same service that also handles the `Identify` protocol described above.
```

{{% notice "note" %}}

> While the blog post mentions that a peer shares it’s dialable addresses with other peers in the
> `Identify` and `Identify Push` messages, what it actually shares is a `Signed Peer Record`
> containing all the addresses and signed with it’s private key.
> That way, anyone who receives that address set can verify that it was announced by the intended peer.
> You can read more about how this works at https://github.com/libp2p/specs/pull/217.

Reviewer: blog post -> page. Link to the RFC itself.
```suggestion
While this page mentions that a peer shares its dialable addresses with other peers in the `Identify` and `Identify Push` messages, what it actually shares is a `Signed Peer Record` containing all the addresses and signed with its private key. That way, anyone who receives that address set can verify that it was indeed created by that peer.
You can read more about how this works in the [Routing Record RFC](https://github.com/libp2p/specs/blob/master/RFC/0003-routing-records.md).
```

---
title: Address Discovery and Advertisement
weight: 2
---

This post explains how a `go-libp2p` node discovers it’s dialable addresses and advertises them to other peers. While it might seem trivial at first glance, there’s a lot going on behind the scenes. Let’s dive in.

### Address Discovery

The following address sets make up a libp2p node’s complete **set of dialable addresses**:

> - The network interface addresses a node explicitly listens on/binds to.

Author: This is done.

#### Network Interface Addresses

- These are the network interface addresses a libp2p node explicitly binds to by either passing in the
`libp2p.ListenAddrs(addrs..)` option when constructing a `Host` or by calling the `h.Network().Listen(addrs..)` API.

- However, the user could pass in an unspecified IPv4 or v6 address here to listen on all available network interfaces.
Thus, we need to resolve the unspecified addresses.
Earlier, we used to resolve the unspecified addresses by using all available network interfaces but this quickly
led to address explosion and a lot of the interfaces were un-usable anyways
(link local interfaces, docker interfaces etc.).

- So, we now resolve unspecified addresses using ONLY the primary network interface addresses
(which we discover using the [netroute][net_route] library) and the loopback interface addresses for both IPv4 and v6.
The primary network interface is basically the interface that would be used to send a packet to the public network
i.e. to an IP address not on the same network as us.
It is usually the “default” entry in the kernel’s Routing table.

#### Publicly visible address(“Observed Addresses”) discovered using Identify

> - If the peer is behind a NAT, the NAT device would assign a NATT’d address to any outgoing packets sent by the peer.

Author: Have addressed both of these valid comments.

      listen addresses.

    - The same address should have been observed by 4 different peers (different here means different IP addresses)
      within the past 40 minutes.

    - The address has been seen at least once in the last 10 minutes(offers some protection against changes in our
      own network conditions).

    - On top of all the above, the observed address manager selects ONLY the “top” two observed addresses for
      each group of (IP address + Transport protocol). For ranking, we prefer observations created because of
      inbound connections over outbound ones(obviously), using the number of peers who reported that same
      observed address to resolve ties.

#### Relay Addresses: Using AutoRelay, AutoNAT and Relay Servers

- We use [circuit relays][spec_relay] to enable other peers to connect to us when we discover we are completely
  NATT’d and are not dialable at all. The basic idea is that we connect to a publicly reachable “relay” server and
  then create a “relay address” which we can advertise.
  A relay address is usually a combination of the Relay server’s address combined with it’s peer id.

>   Any peer that then wants to talk to us can do so via the relay server.

Author: This is done.

- There are three parts to this:
    - Discovering we are completely NATT’d and thus NOT dialable at all using `AutoNAT`.

    - Discovering Relay Servers we can connect to and building relay addresses by using `AutoRelay`.

    - Creating long lived connections with the Relay server and advertising relay addresses in `AutoRelay`.

>     #### AutoNAT: Discovering we are NOT dialable/NATT’d

Reviewer: Formatting looks like it's messed up on these headers due to the indenting. Should probably also be a level 5 header.

Author: Have revisited and fixed all headers.

>     - To discover that we are NOT dialable, we use a component called [AutoNAT][autonat_repo].

Reviewer: That section might be a bit light, but I'd prefer to see the details here reflected there and then just link with a brief overview.
